src/Features/LanguageServer/Protocol/Handler/Rename/RenameHandler.cs

```diff
@@ -64,11 +69,57 @@ public RenameHandler()
                 .GroupBy(docId => renamedSolution.GetRequiredDocument(docId).FilePath, StringComparer.OrdinalIgnoreCase).Select(group => group.First());
 
             var textDiffService = renamedSolution.Workspace.Services.GetRequiredService<IDocumentTextDifferencingService>();
-
             var documentEdits = await ProtocolConversions.ChangedDocumentsToTextDocumentEditsAsync(changedDocuments, renamedSolution.GetRequiredDocument, oldSolution.GetRequiredDocument,
                 textDiffService, cancellationToken).ConfigureAwait(false);
 
-            return new WorkspaceEdit { DocumentChanges = documentEdits };
+            using var _ = ArrayBuilder<SumType<TextDocumentEdit, CreateFile, RenameFile, DeleteFile>>.GetInstance(out var edits);
+
+            // Add the document edits
+            foreach (var documentEdit in documentEdits)
+            {
+                edits.AddRange(documentEdit);
+            }
+
+            // Add an edit for renaming the file (if applicable)
+            if (renameInfo is IInlineRenameInfoWithFileRename renameInfoWithFileRename &&
+                renameInfoWithFileRename.GetFileRenameInfo() == InlineRenameFileRenameInfo.Allowed)
+            {
+                AddFileNameEdit(request.NewName, document, edits);
+            }
+
+            return new WorkspaceEdit { DocumentChanges = edits.ToArray() };
+        }
+
+        private static void AddFileNameEdit(
+            string newName,
+            Document document,
+            ArrayBuilder<SumType<TextDocumentEdit, CreateFile, RenameFile, DeleteFile>> edits)
+        {
+            var fileInfo = new FileInfo(document.FilePath);
+
+            // If a file with the same name already exists within the project, generate a unique name.
+            var uniqueName = GenerateUniqueFileNameWithinProject(document.Project, newName, fileInfo.Extension);
+            var newPath = Path.Combine(fileInfo.DirectoryName, uniqueName);
+
+            edits.Add(new LSP.RenameFile
+            {
+                OldUri = document.GetURI(),
+                NewUri = ProtocolConversions.GetUriFromFilePath(newPath),
+                Options = new LSP.RenameFileOptions { IgnoreIfExists = true }
+            });
+
+            // Local functions
+            static string GenerateUniqueFileNameWithinProject(Project project, string newName, string extension)
+            {
+                var existingFileNames = project.Documents.Select(d => d.Name);
```

Comment:
Is this deliberately an IEnumerable? Seems like a hashset would be a better choice here, or is it working on the assumption that most of the time there won't be collisions?

If so, perhaps a comment would be good.